File: ts-files/main-translated.ts
import { stdout } from "process";
import {
    epoch,
    Grouby as grouby,
    map,
    singleGroup,
    counter,
    filterGroups,
    distinct,
    filter,
    getMappedInt,
    keyGeqInt,
    renameFilterKeys,
    join,
    sumInts,
    dumpAsCsv
} from "./builtins-translated";
import {
    opCreator,
    Operator,
    opResult,
    PacketHeaders,
    $π,
    $$π,
    opResultKind,
} from "./utils-translated";
import { Address4 } from "ip-address";

function ident(end_op: Operator): Operator {
    const firstFun = (headers: PacketHeaders) => {
        const tmpHeaders: Map<string, opResult> = new Map(headers);
        tmpHeaders.delete("eth.src");
        tmpHeaders.delete("eth.dst");
        return tmpHeaders;
    };
    return map(firstFun), end_op;
}

function countPkts(end_op: Operator): opCreator | Operator {
    return $π(epoch(1.0, "eid"),
        $π(grouby(singleGroup, counter, "pkts"),
            end_op
        ));
}

function pktsPerSrcDst(end_op: Operator): opCreator | Operator {
    return $π(epoch(1.0, "eid"),
        $π(grouby(filterGroups(["ipv4.src", "ipv4.dst"]), counter, "pkts"),
            end_op
        ));
}

function distinctSrcs(end_op: Operator): Operator {
    return $π(epoch(1.0, "eid"),
        $π(distinct(filterGroups(["ipv4.src"])),
            $π(grouby(singleGroup, counter, "srcs"),
                end_op
            )));
}

function filterHelper(
    headers: PacketHeaders,
    proto: number,
    flags: number
): boolean {
    return (
        getMappedInt("ipv4.proto", headers) === proto &&
        getMappedInt("l4.flags", headers) === flags
    );
}

function tcpNewCons(end_op: Operator): Operator {
    const threshold: number = 40;
    return $π(epoch(1.0, "eid"),
        $π(filter((headers: PacketHeaders) => filterHelper(headers, 6, 2)),
            $π(grouby(filterGroups(["ipv4.dst"]), counter, "cons"),
                $π(filter(keyGeqInt("cons", threshold)),
                    end_op
                ))));
}

function sshBruteForce(end_op: Operator): Operator {
    const threshold: number = 40;
    return $π(epoch(1.0, "eid"),
        $π(filter((headers: PacketHeaders) => filterHelper(headers, 6, 22)),
            $π(distinct(filterGroups(["ipv4.src", "ipv4.dst", "ipv4.len"])),
                $π(grouby(filterGroups(["ipv4.dst", "ipv4.len"]), counter, "srcs"),
                    $π(filter(keyGeqInt("srcs", threshold)),
                        end_op
                    )))));
}

function superSpreader(end_op: Operator): Operator {
    const threshold: number = 40;
    return $π(epoch(1.0, "eid"),
        $π(distinct(filterGroups(["ipv4.src", "ipv4.dst"])),
            $π(grouby(filterGroups(["ipv4.src"]), counter, "dsts"),
                $π(filter(keyGeqInt("dsts", threshold)),
                    end_op
                ))));
}

function portScan(end_op: Operator): Operator {
    const threshold: number = 40;
    return $π(epoch(1.0, "eid"),
        $π(distinct(filterGroups(["ipv4.src", "l4.dport"])),
            $π(grouby(filterGroups(["ipv4.src"]), counter, "dsts"),
                $π(filter(keyGeqInt("dsts", threshold)),
                    end_op
                ))));
}

function ddos(end_op: Operator): Operator {
    const threshold: number = 45;
    return $π(epoch(1.0, "eid"),
        $π(distinct(filterGroups(["ipv4.src", "ipv4.dst"])),
            $π(grouby(filterGroups(["ipv4.dst"]), counter, "srcs"),
                $π(filter(keyGeqInt("srcs", threshold)),
                    end_op
                ))));
}

function synFloodSonata(
    end_op: Operator,
): Operator[] {
    const threshold: number = 3;
    const epochDur: number = 1.0;

    const syns: opCreator = (next_op: Operator) => {
        return $π(epoch(epochDur, "eid"),
            $π(filter((headers: PacketHeaders) => filterHelper(headers, 6, 22)),
                $π(grouby(filterGroups(["ipv4.dst"]), counter, "syns"),
                    next_op
                )));
    };

    const synacks: opCreator = (next_op: Operator) => {
        return $π(epoch(epochDur, "eid"),
            $π(filter((headers: PacketHeaders) => filterHelper(headers, 6, 18)),
                $π(grouby(filterGroups(["ipv4.src"]), counter, "synacks"),
                    next_op
                )));
    }

    const acks: opCreator = (next_op: Operator) => {
        return $π(epoch(epochDur, "eid"),
            $π(filter((headers: PacketHeaders) => filterHelper(headers, 6, 16)),
                $π(grouby(filterGroups(["ipv4.dst"]), counter, "acks"),
                    next_op
                )));
    }

    const [joinOp1, joinOp2]: [Operator, Operator] =
        $$π(join((headers: PacketHeaders) => {
            return [filterGroups(["host"])(headers),
            filterGroups(["syns+synacks"])(headers)]
        },
            (headers: PacketHeaders) => {
                return [renameFilterKeys([["ipv4.dst", "host"]], headers),
                filterGroups(["acks"])(headers)]
            }), $π(map((headers: PacketHeaders) => {
                return headers.set("syns+synacks",
                    {
                        kind: opResultKind.Int,
                        val: getMappedInt("syns", headers)
                            + getMappedInt("synacks", headers)
                    })
            }),
                $π(filter(keyGeqInt("syns+synacks-acks", threshold)),
                    end_op)));

    const [joinOp3, joinOp4]: [Operator, Operator] =
        $$π(join((headers: PacketHeaders) => {
            return [renameFilterKeys([["ipv4.dst", "host"]], headers),
            filterGroups(["syns"])(headers)]
        }, (headers: PacketHeaders) => {
            return [renameFilterKeys([["ipv4.src", "host"]], headers),
            filterGroups(["synacks"])(headers)]
        }), $π(map((headers: PacketHeaders) => {
            return headers.set("syns+synacks",
                {
                    kind: opResultKind.Int,
                    val: getMappedInt("syns", headers)
                        + getMappedInt("synacks", headers)
                })
        }), joinOp1));

    return [
        syns(joinOp3),
        synacks(joinOp4),
        acks(joinOp2)
    ];
}

function completed_flows(endOp: Operator): Operator[] {
    let threshold: number = 1;
    let epochDur: number = 30.0;

    const syns: opCreator = (nextOp: Operator) => {
        return $π(epoch(epochDur, "eid"),
            $π(filter((headers: PacketHeaders) => filterHelper(headers, 6, 16)),
                $π(grouby(filterGroups(["ipv4.dst"]), counter, "syns"),
                    nextOp
                )));
    }

    const fins: opCreator = (nextOp: Operator) => {
        return $π(epoch(epochDur, "eid"),
            $π(filter((headers: PacketHeaders) => {
                return getMappedInt("ipv4.proto", headers) === 6 &&
                    (getMappedInt("l4.flags", headers) & 1) === 1
            }),
                $π(grouby(filterGroups(["ipv4.src"]), counter, "fins"),
                    nextOp
                )));
    }

    const [joinOp1, joinOp2]: [Operator, Operator] =
        $$π(join((headers: PacketHeaders) => {
            return [renameFilterKeys([["ipv4.dst", "host"]], headers),
            filterGroups(["syns"])(headers)]
        }, (headers: PacketHeaders) => {
            return [renameFilterKeys([["ipv4.src", "host"]], headers),
            filterGroups(["fins"])(headers)]
        }),
            $π(map((headers: PacketHeaders) => {
                return headers.set("diff",
                    {
                        kind: opResultKind.Int,
                        val: getMappedInt("syn", headers)
                            + getMappedInt("fins", headers)
                    })
            }),
                $π(filter(keyGeqInt("diff", threshold)),
                    endOp
                )));

    return [
        syns(joinOp1),
        fins(joinOp2)
    ];
}

function slowloris(end_op: Operator): Operator[] {
    let t1: number = 5;
    let t2: number = 500;
    let t3: number = 90;
    let epochDur: number = 1.0;

    const nConns: opCreator = (nextOp: Operator) => {
        return $π(epoch(epochDur, "eid"),
            $π(filter((headers: PacketHeaders) => getMappedInt("ipv4.proto", headers) === 6),
                $π(distinct(filterGroups(["ipv4.src", "ipv4.dst", "l4.sport"])),
                    $π(grouby(filterGroups(["ipv4.dst"]), counter, "n_conns"),
                        $π(filter((headers: PacketHeaders) => getMappedInt("n_conns", headers) >= t1),
                            nextOp
                        )))));
    }

    const nBytes: opCreator = (nextOp: Operator) => {
        return $π(epoch(epochDur, "eid"),
            $π(filter((headers: PacketHeaders) => getMappedInt("ipv4.proto", headers) === 6),
                $π(grouby(filterGroups(["ipv4.dst"]),
                    (initVal: opResult, header: PacketHeaders) =>
                        sumInts("ipv4.len", initVal, header), "n_bytes"),
                    $π(filter((headers: PacketHeaders) => getMappedInt("n_bytes", headers) >= t2),
                        nextOp
                    ))));
    }

    const [joinOp1, joinOp2]: [Operator, Operator] =
        $$π(join((headers: PacketHeaders) => {
            return [filterGroups(["ipv4.dst"])(headers),
            filterGroups(["n_conns"])(headers)]
        },
            (headers: PacketHeaders) => {
                return [filterGroups(["ipv4.dst"])(headers),
                filterGroups(["n_bytes"])(headers)]
            }),
            $π(map((headers: PacketHeaders) => {
                return headers.set("bytes_per_conn",
                    {
                        kind: opResultKind.Int,
                        val: getMappedInt("n_bytes", headers)
                            + getMappedInt("n_conns", headers)
                    })
            }),
                $π(filter((headers: PacketHeaders) => getMappedInt("bytes_per_conn", headers) <= t3),
                    end_op
                )));

    return [
        nConns(joinOp1),
        nBytes(joinOp2),
    ]
}

function join_test(end_op: Operator): Operator[] {
    let epochDur: number = 1.0;
    const syns: opCreator = (nextOp: Operator) => {
        return $π(epoch(epochDur, "eid"),
            $π(filter((headers: PacketHeaders) => filterHelper(headers, 6, 2)),
                nextOp
            ));
    }

    const synacks: opCreator = (nextOp: Operator) => {
        return $π(epoch(epochDur, "eid"),
            $π(filter((headers: PacketHeaders) => filterHelper(headers, 6, 18)),
                nextOp
            ));
    }

    const [joinOp1, joinOp2]: [Operator, Operator] =
        $$π(join((headers: PacketHeaders) => {
            return [renameFilterKeys([["ipv4.dst", "host"]], headers),
            filterGroups(["syns"])(headers)]
        }, (headers: PacketHeaders) => {
            return [renameFilterKeys([["ipv4.src", "host"]], headers),
            filterGroups(["fins"])(headers)]
        }),
            end_op
        );

    return [
        syns(joinOp1),
        synacks(joinOp2)
    ]
}

function q3(end_op: Operator): Operator {
    return $π(epoch(100.0, "eid"),
        $π(distinct(filterGroups(["ipv4.src", "ipv4.dst"])),
            end_op
        ));
}

function q4(end_op: Operator): Operator {
    return $π(epoch(10000.0, "eid"),
        $π(grouby(filterGroups(["ipv4.dst"]), counter, "pkts"),
            end_op
        ));
}


const queries: Operator[] = [$π(ident, dumpAsCsv(stdout))]

function runQueries(): void {
    let headers: PacketHeaders[] = [];
    for (let i = 1; i <= 20; i++) {
        new Map<string, opResult>([
            ["time", { kind: opResultKind.Float, val: i }],
            ["eth.src", {
                kind: opResultKind.MAC,
                val: new Uint8Array([..."\x00\x11\x22\x33\x44\x55"]
                    .map(c => c.charCodeAt(0)))
            }],
            ["eth.dst", {
                kind: opResultKind.MAC,
                val: new Uint8Array([..."\xAA\xBB\xCC\xDD\xEE\xFF"]
                    .map(c => c.charCodeAt(0)))
            }],
            ["eth.ethertype", { kind: opResultKind.Int, val: 0x0800 }],
            ["ipv4.hlen", { kind: opResultKind.Int, val: 20 }],
            ["ipv4.proto", { kind: opResultKind.Int, val: 6 }],
            ["ipv4.len", { kind: opResultKind.Int, val: 60 }],
            ["ipv4.src", { kind: opResultKind.IPv4, val: new Address4("127.0.0.1") }],
            ["ipv4.dst", { kind: opResultKind.IPv4, val: new Address4("127.0.0.1") }],
            ["l4.sport", { kind: opResultKind.Int, val: 440 }],
            ["l4.dport", { kind: opResultKind.Int, val: 50000 }],
            ["l4.flags", { kind: opResultKind.Int, val: 10 }],
        ])
    }
    headers.forEach((header) => queries.forEach((op) => op.next(header)));
}

runQueries()

File: original-py-translations/main_translated.py
from utils_translated import *
from builtins_translated import *
from functools import partial
import sys
from typing import cast

def ident(next_op: Operator) -> Operator:
    def remove_keys(packet: PacketHeaders) -> PacketHeaders:
        return PacketHeaders({key: val for key, val in packet.items()
                       if key != "eth.src" and key != "eth.dst"})
    return \
    Op_to_op(create_map_operator, remove_keys) \
    >> next_op

def count_pkts(next_op: Operator) -> Operator:
    return \
    Op_to_op(create_epoch_operator, 1.0, "eid") \
    >> (Op_to_op(create_groupby_operator, single_group, counter, "pkts") \
    >> next_op)

def pkts_per_src_dst(next_op: Operator) -> Operator:
    return \
    Op_to_op(create_epoch_operator, 1.0, "eid") \
    >> (Op_to_op(create_groupby_operator,
            partial(filter_groups, ["ipv4.src", "ipv4.dst"]), 
            counter, "pkts") \
    >> next_op)

def distinct_srcs(next_op: Operator) -> Operator:
    return \
    Op_to_op(create_epoch_operator, 1.0, "eid") \
    >> (Op_to_op(create_distinct_operator, 
                partial(filter_groups, ["ipv4.src"])) \
    >> (Op_to_op(create_groupby_operator, single_group, counter, "srcs") \
    >> next_op))

def tcp_new_cons(next_op: Operator) -> Operator:
    threshold: int = 40

    return \
    Op_to_op(create_epoch_operator, 1.0, "eid") \
    >> (Op_to_op(create_filter_operator, partial(filter_helper, 6, 2)) \
    >> (Op_to_op(create_groupby_operator, 
                partial(filter_groups, ["ipv4.dst"]), 
                counter, "cons") \
    >> (Op_to_op(create_filter_operator, 
                partial(key_geq_int, "cons", threshold)) \
    >> next_op)))

def ssh_brute_force(next_op: Operator) -> Operator:
    threshold: int = 40

    return \
    Op_to_op(create_epoch_operator, 1.0, "eid") \
    >> (Op_to_op(create_filter_operator, partial(filter_helper, 6, 22)) \
    >> (Op_to_op(create_distinct_operator, 
                partial(filter_groups, 
                ["ipv4.src", "ipv4.dst", "ipv4.len"])) \
    >> (Op_to_op(create_groupby_operator, 
                partial(filter_groups, ["ipv4.src", "ipv4.dst", "ipv4.len"]),
                counter, "srcs") \
    >> (Op_to_op(create_filter_operator, 
                partial(key_geq_int, "srcs", threshold)) \
    >> next_op))))
    

def super_spreader(next_op: Operator) -> Operator:
    threshold: int = 40
    return \
    Op_to_op(create_epoch_operator, 1.0, "eid") \
    >> (Op_to_op(create_distinct_operator, 
                partial(filter_groups, ["ipv4.src", "l4.dport"])) \
    >> (Op_to_op(create_groupby_operator, 
                partial(filter_groups, ["ipv4.src"]), counter, "ports") \
    >> (Op_to_op(create_filter_operator, 
                partial(key_geq_int, "ports", threshold)) \
    >> next_op)))

def ddos(next_op: Operator) -> Operator:
    threshold: int = 45
    return \
    Op_to_op(create_epoch_operator, 1.0, "eid") \
    >> (Op_to_op(create_distinct_operator, 
                partial(filter_groups, ["ipv4.src", "ipv4.dst"])) \
    >> (Op_to_op(create_groupby_operator, 
                partial(filter_groups, ["ipv4.dst"]), counter, "srcs") \
    >> (Op_to_op(create_filter_operator, 
                partial(key_geq_int, "srcs", threshold)) \
    >> next_op)))

def syn_flood_sonata(next_op: Operator) -> list[Operator]:
    threshold: int = 3
    epoch_dur: float = 1.0

    def syns(end_op: Operator) -> Operator:
        return \
        Op_to_op(create_epoch_operator, epoch_dur, "eid") \
        >> (Op_to_op(create_filter_operator, partial(filter_helper, 6, 2)) \
        >> (Op_to_op(create_groupby_operator, 
                    partial(filter_groups, ["ipv4.dst"]), counter, "syns") \
        >> end_op))
    
    def synacks(end_op: Operator) -> Operator:
        return \
        Op_to_op(create_epoch_operator, epoch_dur, "eid") \
        >> (Op_to_op(create_filter_operator, partial(filter_helper, 6, 18)) \
        >> (Op_to_op(create_groupby_operator, 
                    partial(filter_groups, ["ipv4.dst"]), 
                    counter, "synacks") \
        >> end_op))
    
    def acks(end_op: Operator) -> Operator:
        return \
        Op_to_op(create_epoch_operator, epoch_dur, "eid") \
        >> (Op_to_op(create_filter_operator, partial(filter_helper, 6, 16)) \
        >> (Op_to_op(create_groupby_operator,
                     partial(filter_groups, ["ipv4.dst"]), counter, "acks") \
        >> end_op))
    
    join_ops: tuple[Operator, Operator] = \
        Op_to_op_tup(create_join_operator, 
            lambda packet: 
                (filter_groups(["host"], packet), 
                filter_groups(["syns+synacks"], packet)), 
            lambda packet: 
                (rename_filtered_keys([("ipv4.dst","host")], packet), 
                 filter_groups(["acks"], packet))) \
    >> (Op_to_op(create_map_operator, 
            lambda packet:
                packet.__setitem__("syns+synacks-acks", Int( 
                        packet.get_mapped_int("syns+synacks") - 
                        packet.get_mapped_int("acks")))) \
    >> (Op_to_op(create_filter_operator, 
                partial(key_geq_int, "syns+synacks-acks", threshold)) \
    >> next_op))
    join_op1, join_op2 = join_ops

    join_ops = \
        Op_to_op_tup(create_join_operator,
            lambda packet:
                (rename_filtered_keys([("ipv4.dst","host")], packet), 
                filter_groups(["syns"], packet)), \
            lambda packet:
                (rename_filtered_keys([("ipv4.src","host")], packet), 
                filter_groups(["synacks"], packet))) \
        >> (Op_to_op(create_map_operator, lambda packet:
                packet.__setitem__
                ("syns+synacks", 
                Int( packet.get_mapped_int("syns") + 
                    packet.get_mapped_int("synacks")))) \
        >> join_op1)
    join_op3, join_op4 = join_ops

    return [
            Op_to_op(syns) >> join_op3, 
            Op_to_op(synacks) >> join_op4, 
            Op_to_op(acks) >> join_op2,
           ]

def completed_flows(next_op: Operator) -> list[Operator]:
    threshold: int = 1
    epoch_dur: float = 30.0

    def syns(next_op: Operator) -> Operator:
        return \
        Op_to_op(create_epoch_operator, epoch_dur, "eid") \
        >> (Op_to_op(create_filter_operator, 
            partial(filter_helper, 6, 2)) \
        >> (Op_to_op(create_groupby_operator, 
            partial(filter_groups, ["ipv4.src"]), 
            counter, "syns") \
        >> next_op))
    
    def fins(next_op: Operator) -> Operator:
        return \
        Op_to_op(create_epoch_operator, epoch_dur, "eid") \
        >> (Op_to_op(create_filter_operator, 
                    (lambda packet:
                        packet.get_mapped_int("ipv4.proto") == 6 and \
                        (packet.get_mapped_int("l4.flags") & 1) == 1)) \
        >> (Op_to_op(create_groupby_operator, 
                    partial(filter_groups, ["ipv4.src"]), 
                    counter, "fins") \
        >> next_op))

    operators: tuple[Operator, Operator] = \
        Op_to_op(create_join_operator, 
                (lambda packet:
                    (rename_filtered_keys([("ipv4.dst","host")], packet),
                    filter_groups(["syns"], packet))),
                (lambda packet:
                    (rename_filtered_keys([("ipv4.src","host")], packet),
                    filter_groups(["fins"], packet)))) \
        >> (Op_to_op_tup(create_map_operator,
                    (lambda packet:
                        packet.__setitem__
                        ("diff", Int( 
                        packet.get_mapped_int("syns") - 
                        packet.get_mapped_int("fins"))))) \
        >> (Op_to_op(create_filter_operator, 
            partial(key_geq_int, "diff", threshold)) \
        >> next_op))
    op1, op2 = operators

    return [
            Op_to_op(syns) >> op1,
            Op_to_op(fins) >> op2,
           ]

def slowloris(next_op: Operator) -> list[Operator]:
    t1: int = 5
    t2: int = 500
    t3: int = 90
    epoch_dur: float = 1.0

    def n_conns(next_op: Operator) -> Operator:
        return \
        Op_to_op(create_epoch_operator, epoch_dur, "eid") \
        >> (Op_to_op(create_filter_operator, 
                    (lambda packet: 
                        packet.get_mapped_int("ipv4.proto") == 6)) \
        >> (Op_to_op(create_distinct_operator, 
                    partial(filter_groups, 
                    ["ipv4.src", "ipv4.dst", "l4.sport"])) \
        >> (Op_to_op(create_groupby_operator, 
                    partial(filter_groups, ["ipv4.dst"]), 
                    counter, "n_conns") \
        >> (Op_to_op(create_filter_operator, 
                    (lambda packet: 
                        packet.get_mapped_int("n_conns") >= t1))\
        >> next_op))))

    def n_bytes(next_op: Operator) -> Operator:
        return \
        Op_to_op(create_epoch_operator, epoch_dur, "eid") \
        >> (Op_to_op(create_filter_operator, 
                    (lambda packet: 
                        packet.get_mapped_int("ipv4.proto") == 6)) \
        >> (Op_to_op(create_groupby_operator, 
            partial(filter_groups, ["ipv4.dst"]), 
            partial(sum_ints, "ipv4.len"), "n_bytes") \
        >> (Op_to_op(create_filter_operator, 
                    (lambda packet: 
                        packet.get_mapped_int("n_bytes") >= t2)) \
        >> next_op)))

    operators: tuple[Operator, Operator] = \
        Op_to_op(create_join_operator,
                (lambda packet:
                    (filter_groups(["n_conns"], packet),
                    filter_groups(["n_conns"], packet))),
                (lambda packet:
                    (filter_groups(["ipv4.dst"], packet),
                    filter_groups(["n_bytes"], packet)))) \
        >> (Op_to_op_tup(create_map_operator, 
                        (lambda packet:
                        packet.__setitem__
                        ("bytes_per_conn", Int( 
                        packet.get_mapped_int("n_bytes") - 
                        packet.get_mapped_int("n_conns"))))) \
        >> (Op_to_op(create_filter_operator,
                    (lambda packet:
                        packet.get_mapped_int("bytes_per_conn") <= t3)) \
        >> next_op))
    op1, op2 = operators

    return [
            Op_to_op(n_conns) >> op1,
            Op_to_op(n_bytes) >> op2,
           ]

def create_join_operator_test(next_op: Operator) -> list[Operator]:
    epoch_dur: float = 1.0

    def syns(next_op: Operator) -> Operator:
        return \
        Op_to_op(create_epoch_operator, epoch_dur, "eid") \
        >> (Op_to_op(create_filter_operator, partial(filter_helper, 6, 2)) \
        >> next_op)
    
    def synacks(next_op: Operator) -> Operator:
        return \
        Op_to_op(create_epoch_operator, epoch_dur, "eid") \
        >> (Op_to_op(create_filter_operator, partial(filter_helper, 6, 18)) \
        >> next_op)
    
    operators: tuple[Operator, Operator] = \
        Op_to_op_tup(create_join_operator,
            (lambda packet: (rename_filtered_keys([("ipv4.src","host")], packet), 
                        (rename_filtered_keys([("ipv4.dst","remote")], packet)))),
            (lambda packet: ((rename_filtered_keys([("ipv4.dst","host")], packet)), 
                        (filter_groups(["time"] , packet))))) \
        >> next_op
    op1, op2 = operators
    return [
            Op_to_op(syns) >> op1, 
            Op_to_op(synacks) >> op2
           ]

def q3(next_op: Operator) -> Operator:
    return \
    Op_to_op(create_epoch_operator, 100.0, "eid") \
    >> (Op_to_op(create_distinct_operator, 
                partial(filter_groups, ["ipv4.src", "ipv4.dst"])) \
    >> next_op)

def q4(next_op: Operator) -> Operator:
    return \
    Op_to_op(create_epoch_operator, 10000.0, "eid") \
    >> (Op_to_op(create_groupby_operator, 
                partial(filter_groups, ["ipv4.dst"]), 
                counter, "pkts") \
    >> next_op)

queries: list[Operator] = [q4(dump_as_csv(sys.stdout))]

def run_queries() -> None:           
    [[query.next(packet) for query in queries] for packet in [PacketHeaders({
            "time": Float(0.000000 + cast(float, i)),
            "eth.src": MAC(bytearray(b"\x00\x11\x22\x33\x44\x55")),
            "eth.dst": MAC(bytearray(b"\xAA\xBB\xCC\xDD\xEE\xFF")),
            "eth.ethertype": Int(0x0800),
            "ipv4.hlen": Int(20+i),
            "ipv4.proto": Int(6),
            "ipv4.len": Int(60),
            "ipv4.src": Ipv4(IPv4Address("127.0.0.1")),
            "ipv4.dst": Ipv4(IPv4Address("192.6.8.1")),
            "l4.sport": Int(440),
            "l4.dport": Int(50000),
            "l4.flags": Int(10),
        }) for i in range(4)]]

def main():
    run_queries()
    print("Done")

if __name__ == "__main__":
    main()
